crates/oxc_linter/src/rules/oxc/no_map_spread.rs

```diff
@@ -0,0 +1,545 @@
+        if let AstKind::ReturnStatement(_) = kind {
+            self.is_in_return = false;
+        }
+    }
+
+    fn visit_expression(&mut self, expr: &Expression<'a>) {
+        if !self.is_in_return {
+            return;
+        }
+        match expr.get_inner_expression() {
+            Expression::ObjectExpression(obj) => self.visit_object_expression(obj),
+            Expression::ArrayExpression(arr) => self.visit_array_expression(arr),
+            Expression::Identifier(ident) => {
+                // let Some(reference_id) = ident.reference_id() else {
+                //     return;
+                // };
+                let Some(symbol_id) = ident
+                    .reference_id()
+                    .map(|id| self.ctx.symbols().get_reference(id))
+                    .and_then(Reference::symbol_id)
+                else {
+                    return;
+                };
+                let declaration_scope = self.ctx.symbols().get_scope_id(symbol_id);
+                // symbol is not declared within the mapper callback
+                if !self
+                    .ctx
+                    .scopes()
+                    .ancestors(declaration_scope)
+                    .any(|parent_id| parent_id == self.cb_scope_id)
+                {
+                    return;
+                }
+                let declaration_node =
+                    self.ctx.nodes().get_node(self.ctx.symbols().get_declaration(symbol_id));
+                self.visit_kind(declaration_node.kind());
+            }
+            _ => {}
+        }
+    }
+
+    fn visit_object_expression(&mut self, obj: &ObjectExpression<'a>) {
+        if self.is_in_return
+            && obj
+                .properties
+                .iter()
+                .any(|prop| matches!(prop, ObjectPropertyKind::SpreadProperty(_)))
+        {
+            (self.cb)(Spread::Object(obj));
+        }
+    }
+
+    fn visit_array_expression(&mut self, arr: &ArrayExpression<'a>) {
+        if self.is_in_return
+            && arr
+                .elements
+                .iter()
+                .any(|prop| matches!(prop, ArrayExpressionElement::SpreadElement(_)))
+        {
+            (self.cb)(Spread::Array(arr));
+        }
+    }
+}
+
+#[test]
+fn test() {
+    use crate::tester::Tester;
+    use serde_json::json;
+
+    let pass = vec![
+        ("let a = b.map(x => x)", None),
+        ("let b = []; let a = b.map(x => ({ ...x })); console.log(b)", None),
+        ("let a = b.map(x => Object.assign(x, { foo: { ...x.foo, bar }}))", None),
+        (
+            "class Foo {
+                #arr = []
+                id = 0
+                public get arr() {
+                    return this.#arr.map(x => ({ ...x, id: this.id }))
+                }
+            }",
+            None,
+        ),
+        (
+            // Spread does not occur within `map`
+            "let extras = { ...foo, ...bar }; let a = b.map(x => extras)",
+            None,
+        ),
```

Comment:
i would suggest adding
```ts
function foo(arr) {
    return arr.map(x => ({ ...x }))
}
```

i don't know whether this would be reported currently, but it shuold not be as we don't know whether the fn arg could be used at the calcite

Reply:
sounds good to me

Reply:
I've added `ignoreArgs`. It defaults to `true`.